feat(python): Expand troubleshooting guide (#1796)
getsentry/sentry-docs: src/collections/_documentation/platforms/python/troubleshooting.md

---
title: Troubleshooting
sidebar_order: 11
---

> This section contains various gotchas that are inherent to the Python SDK's

Reviewer: Perhaps " Troubleshooting helps those using the Python SDK with issues inherent to the design; these issues do not necessarily affect most users. Use this section to help answer these questions:"

Author: I've rewritten to be short and direct:

```
We expect most users of the Python SDK not to run into any of the problems documented here.

Use the information in this page to help answer these questions:
```

design, do not necessarily affect the majority of users and do not really fit
anywhere else.

> The information here will help you answer these questions:

Reviewer: delete if you adopt the edit above

but there are quite a few situations where this approach fails.
This section contains various gotchas that are inherent to the Python SDK's
design, do not necessarily affect the majority of users and do not really fit
anywhere else.

Read this document if you cannot figure out why context data is leaking across
your HTTP requests, or when data is missing or popping up at the wrong place
and time.
The information here will help you answer these questions:

> ## Python 2: thread locals and gevent

Reviewer: Use Title case (though do not capitalize technical terms that aren't capitalized:) )

> ## Addressing concurrency issues

Reviewer: Title case here

The Python SDK does its best to figure out how contextual data such as tags set
with `sentry_sdk.set_tags` is supposed to flow along your control flow. In most
cases it works perfectly, but in a few situations some special care must be
taken. This is specially true when working with a code base doing concurrency
outside of the provided framework integrations.

The general recommendation is to have one hub per "concurrency unit"
(thread/coroutine/etc). The SDK ensures every thread has an independent hub. If
you do concurrency with `asyncio` coroutines, clone the current hub for use
within a block that runs concurrent code:

```python
with Hub(Hub.current):
    # in this block Hub.current refers to a new clone
    # of the original hub, with the same client and

>     # the same initial scope data.

Author: @untitaker if you'd be able to help here, I wanted to add a more complete example, similar to what is in getsentry/sentry-python#147 but with your suggested correction applied.

```

Issues with `asyncio` have then an easy workaround: every `asyncio` coroutine
that really does run concurrently with other coroutines needs to be made into a

> task, then the hub needs to be cloned and reassigned.

Reviewer 2: ```suggestion
task, then the hub needs to be cloned as shown above.
```

Author: Thanks. I was not sure what you meant with "reassigned" originally, but now I see that `Hub.__enter__` sets the current hub, so that should be enough.

I wanted to add a more complete example, similar to what is in https://github.com/getsentry/sentry-python/issues/147 but with your suggested correction applied. WDYT?
